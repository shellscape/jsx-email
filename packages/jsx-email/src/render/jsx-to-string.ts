/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';
import hash from 'hash-it';

import { escapeString } from './escape-string';
import { stringifyStyles } from './stringify-styles';

const ATTR_ALIASES = {
  acceptCharset: 'acceptcharset',
  accessKey: 'accesskey',
  allowFullScreen: 'allowfullscreen',
  autoCapitalize: 'autocapitalize',
  autoComplete: 'autocomplete',
  autoCorrect: 'autocorrect',
  autoFocus: 'autofocus',
  autoPlay: 'autoplay',
  charSet: 'charset',
  className: 'class',
  colSpan: 'colspan',
  contentEditable: 'contenteditable',
  crossOrigin: 'crossorigin',
  dateTime: 'datetime',
  defaultChecked: 'checked',
  defaultSelected: 'selected',
  defaultValue: 'value',
  htmlFor: 'for',
  httpEquiv: 'http-equiv',
  longDesc: 'longdesc',
  maxLength: 'maxlength',
  minLength: 'minlength',
  noModule: 'nomodule',
  noValidate: 'novalidate',
  readOnly: 'readonly',
  referrerPolicy: 'referrerpolicy',
  rowSpan: 'rowspan',
  spellCheck: 'spellcheck',
  tabIndex: 'tabindex',
  useMap: 'usemap'
};

// <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>
const BOOLEAN_ATTRS = new Set([
  'async',
  'allowfullscreen',
  'allowpaymentrequest',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'defer',
  'disabled',
  'formnovalidate',
  'hidden',
  'ismap',
  'multiple',
  'muted',
  'novalidate',
  'nowrap',
  'open',
  'readonly',
  'required',
  'reversed',
  'selected'
]);

// https://www.w3.org/TR/html/syntax.html#void-elements
const VOID_ELEMENTS = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr'
]);

const EMPTY_OBJECT = Object.freeze({});

const promiseMap = new Map();

const wrapPromise = <TPromise extends Promise<any>>(promise: TPromise) => {
  let status = 'pending';
  let result: Awaited<TPromise>;
  const suspender = promise.then(
    (r) => {
      status = 'success';
      result = r;
    },
    (e) => {
      status = 'error';
      result = e;
    }
  );
  return {
    // eslint-disable-next-line consistent-return
    read() {
      if (status === 'pending') {
        throw suspender;
      } else if (status === 'error') {
        throw result;
      } else if (status === 'success') {
        return result;
      }
    }
  };
};

export const useData = <TData>(props: any, cb: () => Promise<TData>): TData => {
  const key = hash(props);
  let dataPromise;
  if (promiseMap.has(key)) {
    dataPromise = promiseMap.get(key);
  } else {
    dataPromise = wrapPromise(cb());
    promiseMap.set(key, dataPromise);
  }
  return dataPromise.read();
};

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    throw new Error('Unsupported JSX element:', element as any);
  }

  const { type } = element;
  const props = element.props || EMPTY_OBJECT;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = ATTR_ALIASES[prop as keyof typeof ATTR_ALIASES] || prop;

        if (BOOLEAN_ATTRS.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VOID_ELEMENTS.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      return jsxToString((type as FC)(props));
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    }

    throw new Error(`Unsupported JSX element type: ${String(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
