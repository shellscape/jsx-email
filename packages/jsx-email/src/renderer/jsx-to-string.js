/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */
import chalk from 'chalk';
import { log } from '../log.js';
import { AttributeAliases, BooleanAttributes, EmptyObject, VoidElements } from './constants.js';
import { escapeString } from './escape-string.js';
import { stringifyStyles } from './stringify-styles.js';
import { isValidElement } from './is-valid-element.js';
const renderSuspense = async (children) => {
    try {
        const result = await jsxToString(children);
        return result;
    }
    catch (error) {
        if (error instanceof Promise) {
            await error;
            return renderSuspense(children);
        }
        throw error;
    }
};
/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element) {
    if (element == null) {
        return '';
    }
    if (typeof element === 'string') {
        return escapeString(element);
    }
    else if (typeof element === 'number') {
        return String(element);
    }
    else if (typeof element === 'boolean' || element == null) {
        return '';
    }
    else if (isIterable(element)) {
        let html = '';
        for (const child of element) {
            // eslint-disable-next-line no-await-in-loop
            html += await jsxToString(child);
        }
        return html;
    }
    // Handle Promise case
    if (element instanceof Promise) {
        const resolvedElement = await element;
        return jsxToString(resolvedElement);
    }
    if (!isValidElement(element)) {
        log.error(chalk `{red Unsupported JSX element}:`, element);
        throw new Error(`Unsupported JSX element`);
    }
    const { type } = element;
    const props = element.props || EmptyObject;
    if (typeof type === 'string') {
        let html = `<${type}`;
        let innerHTML = '';
        for (const prop of Object.keys(props)) {
            const value = props[prop];
            if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
                // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
            }
            else if (prop === 'class' || prop === 'className') {
                // This condition is here because it is the most common attribute
                // and short-circuiting results in a ~5% performance boost.
                html += value ? ` class="${escapeString(value)}"` : '';
            }
            else if (prop === 'style') {
                html += ` style="${stringifyStyles(value)}"`;
            }
            else if (prop === 'dangerouslySetInnerHTML') {
                // eslint-disable-next-line no-underscore-dangle
                innerHTML = value.__html;
            }
            else {
                const name = AttributeAliases[prop] || prop;
                if (BooleanAttributes.has(name)) {
                    html += value ? ` ${name}` : '';
                }
                else if (typeof value === 'string') {
                    html += ` ${name}="${escapeString(value)}"`;
                }
                else if (typeof value === 'number') {
                    html += ` ${name}="${String(value)}"`;
                }
                else if (typeof value === 'boolean') {
                    html += ` ${name}="${value}"`;
                }
            }
        }
        if (VoidElements.has(type)) {
            html += '/>';
        }
        else {
            html += '>';
            if (innerHTML) {
                html += innerHTML;
            }
            else {
                html += await jsxToString(props.children);
            }
            html += `</${type}>`;
        }
        return html;
    }
    else if (type) {
        if (typeof type === 'function') {
            const result = type(props);
            const renderedFC = await (result instanceof Promise
                ? result.then(jsxToString)
                : jsxToString(result));
            const sym = type.$$typeof;
            if (sym && Symbol.keyFor(sym) === 'react.provider') {
                type.context.pop();
            }
            return renderedFC;
        }
        else if (typeof type === 'symbol') {
            const key = Symbol.keyFor(type);
            // is this react fragment?
            if (key === 'react.fragment') {
                return jsxToString(props.children);
            }
            else if (key === 'react.suspense') {
                const suspenseResult = await renderSuspense(props.children);
                return suspenseResult;
            }
        }
        else if (isReactForwardRef(type)) {
            return jsxToString(type.render(props, props.ref));
        }
        else if (type.$$typeof) {
            const key = Symbol.keyFor(type.$$typeof);
            if (key === 'react.provider') {
                type.context.pop();
            }
            else if (key === 'react.context') {
                return jsxToString(props.children);
            }
        }
        log.error(chalk `{red Unsupported JSX element}:`, type);
        throw new Error(`Unsupported JSX element type ${JSON.stringify(type)}`);
    }
    return '';
}
function isIterable(node) {
    return typeof node === 'object' && node !== null && Symbol.iterator in node;
}
function isReactForwardRef(type) {
    return (typeof type === 'object' &&
        type !== null &&
        'render' in type &&
        '$$typeof' in type &&
        typeof type.$$typeof === 'symbol' &&
        Symbol.keyFor(type.$$typeof) === 'react.forward_ref');
}
//# sourceMappingURL=jsx-to-string.js.map