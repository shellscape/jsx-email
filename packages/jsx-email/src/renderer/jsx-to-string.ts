/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */
import chalk from 'chalk';
import type { FC, ReactNode } from 'react';

import { log } from '../log.js';

import { AttributeAliases, BooleanAttributes, EmptyObject, VoidElements } from './constants.js';
import { escapeString } from './escape-string.js';
import { stringifyStyles } from './stringify-styles.js';

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    log.error(chalk`{red Unsupported JSX element}:`, element);
    throw new Error(`Unsupported JSX element`);
  }

  const { type } = element;
  const props = element.props || EmptyObject;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = AttributeAliases[prop as keyof typeof AttributeAliases] || prop;

        if (BooleanAttributes.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VoidElements.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      const renderedFC = await jsxToString((type as FC)(props));
      const sym = (type as { $$typeof?: symbol }).$$typeof;

      if (sym && Symbol.keyFor(sym) === 'react.provider') {
        (type as any as { context: any[] }).context.pop();
      }

      return renderedFC;
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    } else if ((type as { $$typeof?: symbol }).$$typeof) {
      const key = Symbol.keyFor((type as { $$typeof: symbol }).$$typeof);
      if (key === 'react.provider') {
        (type as any as { context: any[] }).context.pop();
      } else if (key === 'react.context') {
        return jsxToString(props.children);
      }
    }

    log.error(chalk`{red Unsupported JSX element}:`, type);
    throw new Error(`Unsupported JSX element type ${JSON.stringify(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
