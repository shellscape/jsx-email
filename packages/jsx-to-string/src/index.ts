export * from './jsx-to-string';
