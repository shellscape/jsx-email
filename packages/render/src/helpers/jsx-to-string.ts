/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';

import { escapeString } from './escape-string';
import { stringifyStyles } from './stringify-styles';

const ATTR_ALIASES = {
  acceptCharset: 'acceptcharset',
  accessKey: 'accesskey',
  allowFullScreen: 'allowfullscreen',
  autoCapitalize: 'autocapitalize',
  autoComplete: 'autocomplete',
  autoCorrect: 'autocorrect',
  autoFocus: 'autofocus',
  autoPlay: 'autoplay',
  charSet: 'charset',
  className: 'class',
  colSpan: 'colspan',
  contentEditable: 'contenteditable',
  crossOrigin: 'crossorigin',
  dateTime: 'datetime',
  defaultChecked: 'checked',
  defaultSelected: 'selected',
  defaultValue: 'value',
  htmlFor: 'for',
  httpEquiv: 'http-equiv',
  longDesc: 'longdesc',
  maxLength: 'maxlength',
  minLength: 'minlength',
  noModule: 'nomodule',
  noValidate: 'novalidate',
  readOnly: 'readonly',
  referrerPolicy: 'referrerpolicy',
  rowSpan: 'rowspan',
  spellCheck: 'spellcheck',
  tabIndex: 'tabindex',
  useMap: 'usemap'
};

// <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>
const BOOLEAN_ATTRS = new Set([
  'async',
  'allowfullscreen',
  'allowpaymentrequest',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'defer',
  'disabled',
  'formnovalidate',
  'hidden',
  'ismap',
  'multiple',
  'muted',
  'novalidate',
  'nowrap',
  'open',
  'readonly',
  'required',
  'reversed',
  'selected'
]);

// https://www.w3.org/TR/html/syntax.html#void-elements
const VOID_ELEMENTS = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr'
]);

const EMPTY_OBJECT = Object.freeze({});

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export function jsxToString(element: ReactNode): string {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      html += jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    throw new Error(`Unsupported JSX element: ${String(element)}`);
  }

  const { type } = element;
  const props = element.props || EMPTY_OBJECT;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = ATTR_ALIASES[prop as keyof typeof ATTR_ALIASES] || prop;

        if (BOOLEAN_ATTRS.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VOID_ELEMENTS.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      return jsxToString((type as FC)(props));
    } else if (typeof type === 'symbol') {
      // is this react fragment?
      if (Symbol.keyFor(type) === 'react.fragment') {
        return jsxToString(props.children);
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    }
    throw new Error(`Unsupported JSX element type: ${String(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
