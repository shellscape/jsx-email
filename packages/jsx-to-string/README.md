[npm]: https://img.shields.io/npm/v/@jsx-email/jsx-to-string
[npm-url]: https://www.npmjs.com/package/@jsx-email/jsx-to-string

[![npm][npm]][npm-url]
[![Join our Discord](https://img.shields.io/badge/join_our-Discord-5a64ea)](https://discord.gg/FywZN57mTg)
[![libera manifesto](https://img.shields.io/badge/libera-manifesto-lightgrey.svg)](https://liberamanifesto.com)

<div align="center">
	<img src="https://raw.githubusercontent.com/shellscape/jsx-email/main/assets/npm-header.svg" alt="JSX email"><br/><br/>
</div>

# @jsx-email/jsx-to-string

<div>
  <img src="https://raw.githubusercontent.com/shellscape/jsx-email/main/assets/brackets.svg" alt="JSX email" valign="sub">
  A package for rendering JSX components to HTML strings.
<div><br>

[`@jsx-email/jsx-to-string` Documentation](https://jsx.email/docs/core/jsx-to-string)

## Requirements

This package requires an [LTS](https://github.com/nodejs/Release) Node version (v18.0.0+) and React v18.2.0+.

## Install

```shell
pnpm add @jsx-email/jsx-to-string

# We recommend pnpm - https://pnpm.io
# But npm and yarn are supported
# npm add @jsx-email/jsx-to-string
# yarn add @jsx-email/jsx-to-string
```

## Usage

Convert React components into a HTML string.

```jsx
import { jsxToString } from '@jsx-email/jsx-to-string';

const html = jsxToString(<MyTemplate firstName="Jim" />);
```

Please see the [`@jsx-email/jsx-to-string` Documentation](https://jsx.email/docs/core/jsx-to-string) for props and additional usage information.

## License

[MIT License](./LICENSE.md)
