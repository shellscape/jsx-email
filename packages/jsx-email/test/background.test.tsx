// @ts-ignore
import React from 'react';

import { jsxToString } from '../src/renderer/jsx-to-string.js';
import { Background } from '../src/index.js';

describe('<Hr> component', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();
  });

  it('passes styles and other props correctly', async () => {
    const html = await jsxToString(
      <Background src="url-to-image" width={600} height={400} bgColor="#FFFFFF" bgRepeat="repeat">
        Hello World
      </Background>
    );
    expect(html).toContain('width="600"');
    expect(html).toContain('height="400"');
    expect(html).toContain('height="400"');
  });

  it('renders VML elements conditionally', async () => {
    const html = await jsxToString(
      <Background src="url-to-image" width={600} height={400} bgColor="#FFFFFF" bgRepeat="repeat">
        Hello World
      </Background>
    );
    expect(html).toContain('<!--[if gte mso 9]>');
    expect(html).toContain('v:rect');
    expect(html).toContain('v:fill');
    expect(html).toContain('v:textbox');
    expect(html).toContain('type="tile"');
  });

  it('renders correctly', async () => {
    const actualOutput = await jsxToString(<Background src="url-to-image">Hello World</Background>);
    expect(actualOutput).toMatchSnapshot();
  });
});
