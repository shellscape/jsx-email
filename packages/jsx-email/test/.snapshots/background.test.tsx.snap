// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Hr> component > renders correctly 1`] = `
"<table cellPadding=\\"0\\" cellSpacing=\\"0\\" border=\\"0\\" width=\\"100%\\" role=\\"presentation\\"><tr><td valign=\\"top\\" background=\\"url-to-image\\" style=\\"background-repeat:no-repeat\\"><jsx-email-cond>
<!--[if gte mso 9]>
<v:rect xmlns:v=\\"urn:schemas-microsoft-com:vml\\" fill=\\"true\\" stroke=\\"false\\" style=\\"mso-width-percent:1000;\\"><v:fill type=\\"frame\\" src=\\"url-to-image\\" /><v:textbox style=\\"mso-fit-shape-to-text:true\\" inset=\\"0,0,0,0\\"><![endif]-->
<div>Hello World</div>
<!--[if gte mso 9]></v:textbox></v:rect><![endif]--></jsx-email-cond></td></tr></table>"
`;
